crates/control/src/team_ball_receiver.rs

```diff
@@ -0,0 +1,155 @@
+    received_balls: Players<Option<BallPosition<Field>>>,
+}
+
+#[context]
+pub struct CreationContext {}
+
+#[context]
+pub struct CycleContext {
+    cycle_time: Input<CycleTime, "cycle_time">,
+    filtered_game_controller_state:
+        Input<Option<FilteredGameControllerState>, "filtered_game_controller_state?">,
+    ground_to_field: Input<Option<Isometry2<Ground, Field>>, "ground_to_field?">,
+    network_message: PerceptionInput<Option<IncomingMessage>, "SplNetwork", "filtered_message?">,
+
+    striker_trusts_team_ball: Parameter<Duration, "spl_network.striker_trusts_team_ball">,
+
+    team_balls: AdditionalOutput<Players<Option<BallPosition<Field>>>, "team_balls">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub team_ball: MainOutput<Option<BallPosition<Field>>>,
+    pub network_robot_obstacles: MainOutput<Vec<Point2<Ground>>>,
+}
+
+impl TeamBallReceiver {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            received_balls: Players::default(),
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        let messages = context
+            .network_message
+            .persistent
+            .iter()
+            .flat_map(|(time, messages)| {
+                messages
+                    .iter()
+                    .filter_map(|message| Some((*time, (*message)?)))
+            })
+            .filter_map(|(time, message)| match message {
+                IncomingMessage::Spl(message) => Some((time, *message)),
+                _ => None,
+            });
+        let mut network_robot_obstacles = Vec::new();
+        for (time, message) in messages.clone() {
+            self.process_message(time, message);
+
+            let Some(own_ground_to_field) = context.ground_to_field else {
+                continue;
+            };
+            let pose = match message {
+                HulkMessage::Striker(striker_message) => striker_message.pose,
+                HulkMessage::Loser(loser_message) => loser_message.pose,
+                HulkMessage::VisualReferee(_) => continue,
+            };
+            let sender_position = own_ground_to_field.inverse() * pose.position();
+            network_robot_obstacles.push(sender_position);
+        }
+
+        // Ignore everything during penalty_*
+        if let Some(game_controller_state) = context.filtered_game_controller_state {
+            let in_penalty_shootout = matches!(
+                game_controller_state.game_phase,
+                GamePhase::PenaltyShootout { .. }
+            );
+            let in_penalty_kick = game_controller_state.sub_state == Some(SubState::PenaltyKick);
+
+            if in_penalty_shootout || in_penalty_kick {
+                return Ok(MainOutputs {
+                    team_ball: None.into(),
+                    network_robot_obstacles: Vec::new().into(),
+                });
+            }
+        }
```

Comment:
This can be way earlier at the beginning of the function, can it?

Reply:
Yes, but I'd like to keep updating the internal state during these times. Not important for penalty shootout, but maybe remembering that someone saw a ball a few seconds ago when a penalty kick ends helps.

Reply:
maybe this gets clearer, if we chop the cycle method into even smaller functions, just to improve readability, to easier realize, that you intentionally first update internal state and then return for penalty shootout.
This need is also visible as you wrote comments for structuring the function. These should probably better be functions with descriptive names. 

something like
```
fn cycle(...) {
  self.do_foo();
  let stuff = self.collect_stuff();
  self.update_state(stuff);

  if penalty_shootout {
    return None;
  }

  self.do_bar();
  return Some;
}
```

Reply:
cycle function is much shorter now